### PR TITLE
clientv3: fix disconnect breakage

### PR DESCRIPTION
**Pull request by heyitsanthony**
the etcd-runner lock tester seems to work fine now

### CODE REVIEW

File: clientv3/client.go [hunk 1]
```diff
@@ -50,6 +50,12 @@ type Client struct {
 
 	ctx    context.Context
 	cancel context.CancelFunc
+
+	// reconnc accepts writes which signal the client should reconnect
+	reconnc     chan error
```
Comment (xiang90) [inline clientv3/client.go]: probably comment that the variables are owned by the connMonitor loop.

File: clientv3/client.go [hunk 2]
```diff
@@ -191,7 +202,7 @@ func (c *Client) ActiveConnection() *grpc.ClientConn {
 }
 
 // retryConnection establishes a new connection
-func (c *Client) retryConnection(oldConn *grpc.ClientConn, err error) (*grpc.ClientConn, error) {
+func (c *Client) retryConnection(err error) (newConn *grpc.ClientConn, dialErr error) {
```
Comment (xiang90) [inline clientv3/client.go]: this method name and `connRetry` are too similar I feel. probably we can change the connRetry to requestConnRetry? basically connRetry does not actually retry on the connection but ask the monitor loop to do it.

File: clientv3/client.go [hunk 3]
```diff
@@ -200,24 +211,66 @@ func (c *Client) retryConnection(oldConn *grpc.ClientConn, err error) (*grpc.Cli
...
+	case c.reconnc <- err:
+	default:
+	}
+}
+
+// connWait waits for a reconnect to be processed
+func (c *Client) connWait(ctx context.Context, err error) (*grpc.ClientConn, error) {
+	c.mu.Lock()
+	ch := c.newconnc
+	c.connRetry(err)
```
Comment (xiang90) [inline clientv3/client.go]: line 241 can be outside the lock? i know connRetry is non-blocking. but moving it outside the lock does look better.

File: clientv3/remote.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+package clientv3
+
+import (
+	"sync"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+type remote struct {
```
Comment (xiang90) [inline clientv3/remote.go]: rename this to remoteConn? or remoteClient? we have the remote (API) throughout the codebase. so it is a little bit confusing.

File: clientv3/remote.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	client *Client
+	conn   *grpc.ClientConn
+	update func(*grpc.ClientConn)
```
Comment (xiang90) [inline clientv3/remote.go]: onNewRemoteConn? this is the function we should call when there is a new connection established, right?

**Comment by xiang90** [comment]
Minor issues. Looks good to me in general.

File: clientv3/remote.go (hunk 5 above)
Comment (heyitsanthony) [inline clientv3/remote.go]: `updateConn`? It won't necessarily catch every new remote connection so `on` seems misleading.

Comment (xiang90) [inline clientv3/remote.go]: ok

**Comment by xiang90** [comment]
squash and merge?

**Comment by heyitsanthony** [comment]
ok will do

**Pull request merged**